extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/QuarkusFaultToleranceOperationProvider.java

```diff
@@ -1,22 +1,76 @@
 package io.quarkus.smallrye.faulttolerance.runtime;
 
 import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
 
 import javax.annotation.Priority;
-import javax.enterprise.context.Dependent;
 import javax.enterprise.inject.Alternative;
+import javax.inject.Singleton;
+
+import org.jboss.logging.Logger;
 
 import io.smallrye.faulttolerance.FaultToleranceOperationProvider;
 import io.smallrye.faulttolerance.config.FaultToleranceOperation;
 
-@Dependent
+@Singleton
 @Alternative
 @Priority(1)
 public class QuarkusFaultToleranceOperationProvider implements FaultToleranceOperationProvider {
+    private static final Logger LOG = Logger.getLogger(QuarkusFaultToleranceOperationProvider.class);
+
+    private final Map<CacheKey, FaultToleranceOperation> operationCache = new ConcurrentHashMap<>();
+    private final Function<CacheKey, FaultToleranceOperation> cacheFunction = new Function<CacheKey, FaultToleranceOperation>() {
+        @Override
+        public FaultToleranceOperation apply(CacheKey key) {
+            return createAtRuntime(key);
+        }
+    };
+
+    /**
+     * Called by SmallryeFaultToleranceRecorder to init the operation cache.
+     */
+    void init(Map<CacheKey, FaultToleranceOperation> operationCache) {
+        this.operationCache.putAll(operationCache);
+    }
 
     @Override
     public FaultToleranceOperation get(Class<?> beanClass, Method method) {
-        return FaultToleranceOperation.of(beanClass, method);
+        CacheKey key = new CacheKey(beanClass, method);
+        FaultToleranceOperation existing = operationCache.get(key);
+        return existing != null ? existing : operationCache.computeIfAbsent(key, cacheFunction);
+    }
+
+    private FaultToleranceOperation createAtRuntime(CacheKey key) {
+        LOG.debugf("FaultToleranceOperation not found in the cache for %s creating it at runtime", key);
+        return FaultToleranceOperation.of(key.beanClass, key.method);
     }
 
+    static class CacheKey {
+        private Class<?> beanClass;
+        private Method method;
+
+        public CacheKey(Class<?> beanClass, Method method) {
+            this.beanClass = beanClass;
+            this.method = method;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
+            CacheKey cacheKey = (CacheKey) o;
+            return Objects.equals(beanClass, cacheKey.beanClass) &&
+                    Objects.equals(method, cacheKey.method);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(beanClass, method);
```

Comment:
It's minor but I would compute the hashCode in the constructor if it can become a hot path. That's what we do in HV.

That allows to cache the hashCode once and for all for all the items that are in the map.

Reply:
+1 and also make the fields final please.

Reply:
Actually, I'm torn on that, not having the fields final make the JDK do some weird optimizations :). I removed them from the key objects in HV after Sanne gave me the idea to test it and benchmark it.

Reply:
So I'm not following. Was it better or worse ;-). In any case, the key should be immutable and it's not.

Reply:
@gmset good catch on the cache of the hashCode, I'm not used to do it but totaly worth it :)

Regarding final fields, I add final to all three fields, I can remove it.

@gmset it's very interesting, I was not aware of performance penalty on final field writes inside a constructor so I make some research and yes, there is a small performance penalty because the VM emmit a memory barrier when one of the field written inside a constructor is final. The cost of the barrier is minimal (5ns on ARM, almost nefligeable on x86). As always some good explaination/experiments on this can be found on Alexey Shipilev blog: https://shipilev.net/blog/2014/all-fields-are-final/

I you want, I can remove the three final modifier ...

Reply:
In my HV benchmarks (and I have no idea why), removing the final from the fields did help. In HV, the only classes that don't have final fields are the classes used as keys. It's a very hot path though.

Reply:
OK, to please @gsmet ;) I remove the final modifier.
Ready to be merged